### PR TITLE
[CALCITE-3590] Support SQL hints for Aggregate

### PR DESCRIPTION
**Pull request by cshuo**
Currently in [CALCITE-482](https://issues.apache.org/jira/browse/CALCITE-482), hints are only supported for Project, Join and TableScan. It's also necessary to support hints for Aggregate so as to enable fine grained Aggregate configuration, e.g., two-phase aggregation.
```
SELECT /*+ AGG_STRATEGY(TWO_PHASE) */ 
dep_id,
grade, 
avg(annual_salary),
min(annual_salary),
max(annual_salary)
FROM T
GROUP BY dep_id, grade;
```

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/hint/HintStrategyOr.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Strategy to decide if a hint should apply to a relational expression
+ * by using a series of {@link HintStrategy} rules in a given order.
+ * Returns true if any rule satisfies, otherwise returns false.
+ */
+public class HintStrategyOr implements HintStrategy {
+  //~ Instance fields --------------------------------------------------------
+
+  private ImmutableList<HintStrategy> strategies;
+
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/hint/HintStrategyOr.java]: We can merge this class with `HintStrategyCascade`, you can reference `CompositeOperandTypeChecker`.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 2]
```diff
@@ -704,12 +704,12 @@ protected void convertSelectImpl(
           .accept(
               new RelShuttleImpl() {
                 boolean attached = false;
-                @Override public RelNode visit(LogicalProject project) {
-                  if (!attached) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java]: Change the comment also.

File: core/src/main/java/org/apache/calcite/rel/hint/HintStrategyOr.java (hunk 1 above)
Comment (cshuo) [reply core/src/main/java/org/apache/calcite/rel/hint/HintStrategyOr.java]: Make sense.  👍

File: core/src/main/java/org/apache/calcite/rel/core/Aggregate.java [hunk 3]
```diff
@@ -199,9 +227,9 @@ private boolean isPredicate(RelNode input, int index) {
    * Creates an Aggregate by parsing serialized output.
    */
   protected Aggregate(RelInput input) {
-    this(input.getCluster(), input.getTraitSet(), input.getInput(),
-        input.getBitSet("group"), input.getBitSetList("groups"),
-        input.getAggregateCalls("aggs"));
+    this(input.getCluster(), input.getTraitSet(), new ArrayList<>(),
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/rel/core/Aggregate.java]: nit: As the internal data structure is Immutable, you can use `Collections.emptyList()` to create an ImmutableList directly that  using `new ArrayList<>()`.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/core/Aggregate.java]: Thanks for the review @amaliujia , i missed you comments when i commit the code, sorry for that.

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/rel/core/Aggregate.java]: no worries. It's just a `nit`. Thanks for this contribution and quickly review. I can build hot key hint on top of it now.

**Pull request closed without merging**